helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -37,14 +39,23 @@
     NODES_CONTAIN_SAME_PARTITION,
   }
 
+  public HardConstraint(Type failureReason) {
+    _type = failureReason;
+  }
+
+  public String getFailureReason() {
+    // by default just return the type, extend the method if more details are needed
+    return _type.toString();
+  }
+
   /**
-   * @return True if the proposed assignment is valid.
+   * Check if the replica could be assigned to the node
+   * @return True if the proposed assignment is valid; False otherwise
    */
-  boolean isAssignmentValid(AssignableNode node, AssignableReplica rep,
+  public abstract boolean isAssignmentValid(AssignableNode node, AssignableReplica replica,
```

Comment:
Why we need it to be public here? Can we just put the algorithm implementation class in the same package?